### PR TITLE
Fix ExecutionContext and PartitionTable bugs caught by enabling parallel updates

### PR DESCRIPTION
**Pull request by nbauernfeind**

### CODE REVIEW

File: py/server/tests/test_partitioned_table.py [hunk 1]
```diff
@@ -116,6 +117,12 @@ def test_constituents(self):
         self.assertGreater(len(constituent_tables), 0)
 
     def test_transform(self):
+        _JExecutionContext = jpy.get_type("io.deephaven.engine.context.ExecutionContext")
+        context = _JExecutionContext.newBuilder() \
+                .captureCompilerContext()         \
+                .captureQueryLibrary()            \
+                .emptyQueryScope()                \
+                .build().open()
```
Comment (jmao-denver) [inline py/server/tests/test_partitioned_table.py:125]: This puts too much burden on the users. The ExecutionContext should be transparent to the application code. We might need to redesign it or find a way in Python to wrap it.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/sources/CharacterSparseArraySource.java [hunk 2]
```diff
@@ -414,38 +414,39 @@ public void ensurePrevious(RowSet changedRows) {
...
+            prevFlusher.maybeActivate();
         }
-        prevFlusher.maybeActivate();
 
-        try (final RowSet.Iterator it = changedRows.iterator()) {
-            long key = it.nextLong();
-            while (true) {
-                final long firstKey = key;
+        try (final RowSequence.Iterator it = changedRows.getRowSequenceIterator()) {
+            while (it.hasMore()) {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/sources/CharacterSparseArraySource.java:424]: This could be a `do-while` loop, since we check `changedRows.isEmpty()` above.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/sources/CharacterSparseArraySource.java [hunk 3]
```diff
@@ -414,38 +414,39 @@ public void ensurePrevious(RowSet changedRows) {
             return;
         }
 
-        if (prevFlusher == null) {
-            return;
+        // Note that this method is used to precondition the column source before processing an update in parallel. So,
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/sources/CharacterSparseArraySource.java:417]: Maybe we should rename the method. `ensurePreviousAndPreallocateBlocks` or some other such mouthful.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/QueryTable.java [hunk 4]
```diff
@@ -1266,7 +1267,10 @@ private Table selectOrUpdate(Flavor flavor, final SelectColumn... selectColumns)
 
                     final QueryTable resultTable;
                     final LivenessScope liveResultCapture = isRefreshing() ? new LivenessScope() : null;
-                    try (final SafeCloseable ignored = liveResultCapture != null ? liveResultCapture::release : null) {
+                    try (final SafeCloseable ignored1 = liveResultCapture != null ? liveResultCapture::release : null;
+                            final SafeCloseable ignored2 = ExecutionContext.getDefaultContext().open()) {
+                        // we open the default context here to ensure that the update processing happens in the default
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/QueryTable.java:1272]: This is to save the user from himself. If the formula column needs access to the ExecutionContext it should be installed by the user from within his formula. By opening the default poisoned context, we guarantee that the `fakeUpdate` pass through the formula is invoked without any user installed context nor any systemic context.

Note that the context installed at this point in time is not the context that will be installed from a view column (maybe we should install the default poisoned execution context inside of ConstructSnapshot, too). As discussed in chat, it might be appropriate to back-out the change for update graph notifications to specify their execution context; the user is supposed to install the execution context from within their formula.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/SelectOrUpdateListener.java [hunk 5]
```diff
@@ -103,6 +103,7 @@ protected void onError(Exception error) {
 
     private void handleException(Exception e) {
         dependent.notifyListenersOnError(e, getEntry());
+        getParent().removeUpdateListener(this);
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/SelectOrUpdateListener.java]: One side effect of this approach is that each failing listener pass this same error to the `AsyncClientErrorNotifier`. This simple case prints out four such errors:
```groovy
import io.deephaven.engine.util.TableTools

err = { _ -> if (true) throw new RuntimeException() else 0 }
tt = TableTools.timeTable("00:00:01").update("I = ii", "J = new Object()")
m = TableTools.merge(tt, tt.update("I = ii", "J = I > 5 ? err() : 0"))
```

File: engine/table/src/main/java/io/deephaven/engine/table/impl/InstrumentedTableListenerBase.java [hunk 6]
```diff
@@ -152,9 +152,12 @@ public boolean satisfied(final long step) {
 
     @Override
     public void onFailure(Throwable originalException, Entry sourceEntry) {
+        deregisterOnFailure();
         onFailureInternal(originalException, sourceEntry == null ? entry : sourceEntry);
     }
 
+    protected abstract void deregisterOnFailure();
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/InstrumentedTableListenerBase.java:159]: I think this current implementation is a big improvement over the status quo. I'm trying to think if we could go one better and drop the reference count to zero, thus forcing a destroy. Maybe this whole change set belongs in a separate PR.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/partitioned/PartitionedTableProxyImpl.java [hunk 7]
```diff
@@ -128,20 +128,29 @@ public boolean sanityChecksJoins() {
...
-            context = ExecutionContext.newBuilder()
+            final ExecutionContext.Builder builder = ExecutionContext.newBuilder()
                     .captureCompilerContext()
-                    .markSystemic()
-                    .build();
+                    .markSystemic();
+            if (requiresFullContext) {
+                builder.newQueryLibrary();
+                builder.emptyQueryScope();
+            }
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/partitioned/PartitionedTableProxyImpl.java:140]: This is for `AggFormula`?

File: engine/table/src/main/java/io/deephaven/engine/table/impl/SelectOrUpdateListener.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/SelectOrUpdateListener.java]: This is now pulled out into #2789 and evolving separately.

**Review by rcaudy — APPROVED** [review]
